go/vt/vtgate/safe_session.go

```diff
@@ -144,14 +186,28 @@ func (session *SafeSession) Append(shardSession *vtgatepb.Session_ShardSession,
 	defer session.mu.Unlock()
 
 	if session.autocommitState == autocommitted {
-		panic("BUG: SafeSession.Append: unexpected autocommit state")
+		// Unreachable.
+		return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "BUG: SafeSession.Append: unexpected autocommit state")
+	}
+	if !session.Session.InTransaction {
+		// Unreachable.
+		return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "BUG: SafeSession.Append: not in transaction")
 	}
+	session.autocommitState = notAutocommittable
 
 	// Always append, in order for rollback to succeed.
-	session.ShardSessions = append(session.ShardSessions, shardSession)
-	if session.isSingleDB(txMode) && len(session.ShardSessions) > 1 {
-		session.mustRollback = true
-		return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "multi-db transaction attempted: %v", session.ShardSessions)
+	switch session.commitOrder {
+	case commitOrderNormal:
+		session.ShardSessions = append(session.ShardSessions, shardSession)
+		// isSingle is enforced only for normmal commit order operations.
+		if session.isSingleDB(txMode) && len(session.ShardSessions) > 1 {
+			session.mustRollback = true
+			return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "multi-db transaction attempted: %v", session.ShardSessions)
+		}
+	case commitOrderPre:
+		session.PreSessions = append(session.PreSessions, shardSession)
+	case commitOrderPost:
+		session.PostSessions = append(session.PostSessions, shardSession)
```

Comment:
why isn't this:
`		session.PostSessions = append(shardSession, session.PostSessions)`

Reply:
The three session variables are completely independent of each other. You could potentially have two transactions going to the same shard, one from each pool. If so, the ones in the `Pre` session are committed first, and the ones from `Post` are committed last.